{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bbe500f7-0396-450a-9e0d-48d9620ef6c4",
   "metadata": {},
   "source": [
    "# __STRATEGIE D'ANALYSE BIOINFORMATIQUE - MENTORING PROJECT__\n",
    "\n",
    "__AGMIMONHAN Attolou Raoul, NAME Pakyendou Estel__\n",
    "\n",
    "__Tuteurs: Aurore COMTE & Sebastien RAVEL__\n",
    "\n",
    "Jupyter inspired by the model created by C. Tranchant (DIADE-IRD), J. Orjuela (DIADE-IRD), F. Sabot (DIADE-IRD) and A. Dereeper (PHIM-IRD)\n",
    "***\n",
    "\n",
    "# <span style=\"color: #006E7F\">Table of contents</span>\n",
    "<a class=\"anchor\" id=\"home\"></a>\n",
    "\n",
    "\n",
    "[PRACTICE VI REPRISE- PCA - DAPC - OTHERS](#Genetic_dversity)\n",
    "\n",
    "   \n",
    "   \n",
    "\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dd25e47-827d-4dd7-82a8-cea71fb739a9",
   "metadata": {},
   "source": [
    "# __Practice VI__"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf3a4ef9-cdd7-4030-a73f-faf7d451b07b",
   "metadata": {},
   "source": [
    "## Evaluate some statistiques on VCF file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eee86937-e7b8-4b51-99f1-cbd79e1a8bd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Se déplacer dans le répertoire créé\n",
    "\n",
    "cd /scratch/MOryzae/SCRIPTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acf6d961-5f76-461f-a82f-3ef04d5b25ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Ouvrir l'éditeur de texte nano\n",
    "\n",
    "nano snp2_statistic.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9299fd4-4dac-4e0f-b0cf-61cd31e649c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############ SLURM Configuration ##############\n",
    "\n",
    "### Define Job name\n",
    "#SBATCH --job-name=snp_statistics\n",
    "\n",
    "### Define partition to use\n",
    "#SBATCH -p normal\n",
    "\n",
    "### Define number of CPUs to use\n",
    "#SBATCH -c 8\n",
    "\n",
    "### Specify the node to run on\n",
    "#SBATCH --nodelist=node20  # Run the job on node20\n",
    "\n",
    "#################################################\n",
    "\n",
    "########### Execution Commands ###################\n",
    "\n",
    "# Define directories\n",
    "SUBSET_VCF=\"/scratch/MOryzae/SNP2/vcf_files/snp_correct.vcf.gz\"\n",
    "OUT_DIR=\"/scratch/MOryzae/SNP2/Others_stats\"\n",
    "\n",
    "# Create output directory if it doesn't exist\n",
    "mkdir -p \"$OUT_DIR\"\n",
    "\n",
    "# Load necessary modules\n",
    "module load vcftools/0.1.16\n",
    "\n",
    "# Define base output name\n",
    "OUT_PREFIX=\"${OUT_DIR}/output\"\n",
    "\n",
    "# Calculate allele frequency\n",
    "vcftools --gzvcf \"$SUBSET_VCF\" --remove-indels --freq2 --max-alleles 2 --out \"$OUT_PREFIX\"\n",
    "\n",
    "# Calculate mean depth per individual\n",
    "vcftools --gzvcf \"$SUBSET_VCF\" --remove-indels --depth --out \"$OUT_PREFIX\"\n",
    "\n",
    "# Calculate mean depth per site\n",
    "vcftools --gzvcf \"$SUBSET_VCF\" --remove-indels --site-mean-depth --out \"$OUT_PREFIX\"\n",
    "\n",
    "# Calculate site quality\n",
    "vcftools --gzvcf \"$SUBSET_VCF\" --remove-indels --site-quality --out \"$OUT_PREFIX\"\n",
    "\n",
    "# Calculate proportion of missing data per individual\n",
    "vcftools --gzvcf \"$SUBSET_VCF\" --remove-indels --missing-indv --out \"$OUT_PREFIX\"\n",
    "\n",
    "# Calculate proportion of missing data per site\n",
    "vcftools --gzvcf \"$SUBSET_VCF\" --remove-indels --missing-site --out \"$OUT_PREFIX\"\n",
    "\n",
    "# Calculate heterozygosity and inbreeding coefficient per individual\n",
    "vcftools --gzvcf \"$SUBSET_VCF\" --remove-indels --het --out \"$OUT_PREFIX\"\n",
    "\n",
    "echo \"All SNP statistics have been computed. Results saved to $OUT_DIR\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73e05259-007e-4c9a-a282-f9e736405fae",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Lancer le script\n",
    "\n",
    "sbash snp2_statistic.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5497debe-4200-4acd-89ce-8862f016959e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Déplacer tout le répertoireOthers_stats vers le NAS\n",
    "\n",
    "scp -r /scratch/MOryzae/SNP2/Others_stats san:/projects/medium/CIBiG_MOryzae/SNP2/\n",
    "\n",
    "scp -r /scratch/MOryzae/SNP2/stats san:/projects/medium/CIBiG_MOryzae/SNP2/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6a2cc3a-1edf-4183-82e6-ff044e61a015",
   "metadata": {},
   "outputs": [],
   "source": [
    "Analyser les résultats de statistiques\n",
    "Et définir les paramètres de filtre des SNPs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53110813-62bc-41b0-8682-71dcb67437e4",
   "metadata": {},
   "source": [
    "## __Examining statistics in R__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "509bf435-2acb-464e-a2b6-5adbaf135260",
   "metadata": {},
   "outputs": [],
   "source": [
    "Lancer les analyses statistiques sur sa machine locale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32b6f71d-cb5f-4f4e-aee2-a68e7bcfbf1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load tidyverse package\n",
    "\n",
    "library(tidyverse)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8335a1f3-d88a-41ba-93ae-99b6620e55b4",
   "metadata": {},
   "source": [
    "## Ouvrir l'éditeur de texte nano\n",
    "\n",
    "nano snp_statistic.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bccd2c3d-413e-4679-848b-7d7861eaa05b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load necessary libraries\n",
    "library(tidyverse)\n",
    "library(ggplot2)\n",
    "\n",
    "# Define working directories\n",
    "input_dir <- \"/home/name/Documents/Projet_CIBiG/Mentoring_Project/Results/SNP2/Others_stats\"\n",
    "output_dir <- \"/home/name/Documents/Projet_CIBiG/Mentoring_Project/Results/SNP2/Others_stats/Plots\"\n",
    "summary_file <- file.path(output_dir, \"summary_statistics.txt\")\n",
    "\n",
    "# Create output directory if it doesn't exist\n",
    "if (!dir.exists(output_dir)) {\n",
    "  dir.create(output_dir, recursive = TRUE)\n",
    "}\n",
    "\n",
    "# Set working directory\n",
    "setwd(input_dir)\n",
    "\n",
    "# Open summary file for writing\n",
    "summary_conn <- file(summary_file, open = \"w\")\n",
    "\n",
    "# Variant-based statistics\n",
    "\n",
    "# Variant quality\n",
    "var_qual <- read_delim(file.path(input_dir, \"output.lqual\"), delim = \"\\t\",\n",
    "                       col_names = c(\"chr\", \"pos\", \"qual\"), skip = 1)\n",
    "\n",
    "p <- ggplot(var_qual, aes(qual)) +\n",
    "  geom_density(fill = \"dodgerblue1\", colour = \"black\", alpha = 0.3) +\n",
    "  theme_light() +\n",
    "  ggtitle(\"Variant Quality Distribution\")\n",
    "ggsave(file.path(output_dir, \"variant_quality_density.png\"), plot = p)\n",
    "\n",
    "summary_var_qual <- summary(var_qual$qual)\n",
    "writeLines(\"### Variant Quality Summary ###\\n\", summary_conn)\n",
    "writeLines(capture.output(summary_var_qual), summary_conn)\n",
    "\n",
    "# Variant mean depth\n",
    "var_depth <- read_delim(file.path(input_dir, \"output.ldepth.mean\"), delim = \"\\t\",\n",
    "                        col_names = c(\"chr\", \"pos\", \"mean_depth\", \"var_depth\"), skip = 1)\n",
    "\n",
    "p <- ggplot(var_depth, aes(mean_depth)) +\n",
    "  geom_density(fill = \"dodgerblue1\", colour = \"black\", alpha = 0.3) +\n",
    "  theme_light() +\n",
    "  xlim(0, 100) +\n",
    "  ggtitle(\"Mean Depth per Variant\")\n",
    "ggsave(file.path(output_dir, \"variant_depth_density.png\"), plot = p)\n",
    "\n",
    "summary_var_depth <- summary(var_depth$mean_depth)\n",
    "writeLines(\"\\n### Variant Mean Depth Summary ###\\n\", summary_conn)\n",
    "writeLines(capture.output(summary_var_depth), summary_conn)\n",
    "\n",
    "# Variant missingness\n",
    "var_miss <- read_delim(file.path(input_dir, \"output.lmiss\"), delim = \"\\t\",\n",
    "                       col_names = c(\"chr\", \"pos\", \"nchr\", \"nfiltered\", \"nmiss\", \"fmiss\"), skip = 1)\n",
    "\n",
    "p <- ggplot(var_miss, aes(fmiss)) +\n",
    "  geom_density(fill = \"dodgerblue1\", colour = \"black\", alpha = 0.3) +\n",
    "  theme_light() +\n",
    "  ggtitle(\"Missingness per Variant\")\n",
    "ggsave(file.path(output_dir, \"variant_missingness_density.png\"), plot = p)\n",
    "\n",
    "summary_var_miss <- summary(var_miss$fmiss)\n",
    "writeLines(\"\\n### Variant Missingness Summary ###\\n\", summary_conn)\n",
    "writeLines(capture.output(summary_var_miss), summary_conn)\n",
    "\n",
    "# Minor allele frequency\n",
    "var_freq <- read_delim(file.path(input_dir, \"output.frq\"), delim = \"\\t\",\n",
    "                       col_names = c(\"chr\", \"pos\", \"nalleles\", \"nchr\", \"a1\", \"a2\"), skip = 1)\n",
    "\n",
    "# Calculate minor allele frequency\n",
    "var_freq <- var_freq %>%\n",
    "  mutate(maf = pmin(as.numeric(a1), as.numeric(a2), na.rm = TRUE))\n",
    "\n",
    "p <- ggplot(var_freq, aes(maf)) +\n",
    "  geom_density(fill = \"dodgerblue1\", colour = \"black\", alpha = 0.3) +\n",
    "  theme_light() +\n",
    "  ggtitle(\"Minor Allele Frequency Distribution\")\n",
    "ggsave(file.path(output_dir, \"maf_density.png\"), plot = p)\n",
    "\n",
    "summary_var_freq <- summary(var_freq$maf)\n",
    "writeLines(\"\\n### Minor Allele Frequency Summary ###\\n\", summary_conn)\n",
    "writeLines(capture.output(summary_var_freq), summary_conn)\n",
    "\n",
    "# Individual-based statistics\n",
    "\n",
    "# Mean depth per individual\n",
    "ind_depth <- read_delim(file.path(input_dir, \"output.idepth\"), delim = \"\\t\",\n",
    "                        col_names = c(\"ind\", \"nsites\", \"depth\"), skip = 1)\n",
    "\n",
    "p <- ggplot(ind_depth, aes(depth)) +\n",
    "  geom_histogram(fill = \"dodgerblue1\", colour = \"black\", alpha = 0.3) +\n",
    "  theme_light() +\n",
    "  ggtitle(\"Mean Depth per Individual\")\n",
    "ggsave(file.path(output_dir, \"individual_depth_histogram.png\"), plot = p)\n",
    "\n",
    "summary_ind_depth <- summary(ind_depth$depth)\n",
    "writeLines(\"\\n### Individual Mean Depth Summary ###\\n\", summary_conn)\n",
    "writeLines(capture.output(summary_ind_depth), summary_conn)\n",
    "\n",
    "# Missing data per individual\n",
    "ind_miss <- read_delim(file.path(input_dir, \"output.imiss\"), delim = \"\\t\",\n",
    "                       col_names = c(\"ind\", \"ndata\", \"nfiltered\", \"nmiss\", \"fmiss\"), skip = 1)\n",
    "\n",
    "p <- ggplot(ind_miss, aes(fmiss)) +\n",
    "  geom_histogram(fill = \"dodgerblue1\", colour = \"black\", alpha = 0.3) +\n",
    "  theme_light() +\n",
    "  ggtitle(\"Missing Data per Individual\")\n",
    "ggsave(file.path(output_dir, \"individual_missing_data_histogram.png\"), plot = p)\n",
    "\n",
    "summary_ind_miss <- summary(ind_miss$fmiss)\n",
    "writeLines(\"\\n### Individual Missing Data Summary ###\\n\", summary_conn)\n",
    "writeLines(capture.output(summary_ind_miss), summary_conn)\n",
    "\n",
    "# Heterozygosity and inbreeding coefficient\n",
    "ind_het <- read_delim(file.path(input_dir, \"output.het\"), delim = \"\\t\",\n",
    "                      col_names = c(\"ind\", \"ho\", \"he\", \"nsites\", \"f\"), skip = 1)\n",
    "\n",
    "p <- ggplot(ind_het, aes(f)) +\n",
    "  geom_histogram(fill = \"dodgerblue1\", colour = \"black\", alpha = 0.3) +\n",
    "  theme_light() +\n",
    "  ggtitle(\"Inbreeding Coefficient Distribution\")\n",
    "ggsave(file.path(output_dir, \"inbreeding_coefficient_histogram.png\"), plot = p)\n",
    "\n",
    "summary_ind_het <- summary(ind_het$f)\n",
    "writeLines(\"\\n### Inbreeding Coefficient Summary ###\\n\", summary_conn)\n",
    "writeLines(capture.output(summary_ind_het), summary_conn)\n",
    "\n",
    "# Close summary file\n",
    "close(summary_conn)\n",
    "\n",
    "message(\"All plots and summary statistics have been saved to \", output_dir)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fe64bb0-1259-414f-885e-6f33dba2c104",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Lancer le scriptdans R\n",
    "\n",
    "source(\"/path/to/working dorectory/on your laptop/snp2_statistic.R\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b2bbd97-c75b-4c77-b39c-87764c679875",
   "metadata": {},
   "source": [
    "## __String mode SNP filtering__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b4bdb05-e6e8-491a-89fc-43ee2fd215f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Se déplacer dans le répertoire créé\n",
    "\n",
    "cd /scratch/MOryzae/SCRIPTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "889aa649-759e-445b-beb9-6bd10886f756",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Ouvrir l'éditeur de texte nano\n",
    "\n",
    "nano snp2_filtering.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eea2bd9c-0165-45ae-8153-dafa5b013e09",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############ SLURM Configuration ##############\n",
    "\n",
    "### Define Job name\n",
    "#SBATCH --job-name=snp_filtering\n",
    "#SBATCH -p normal\n",
    "#SBATCH -c 8\n",
    "#SBATCH --nodelist=node20\n",
    "\n",
    "#################################################\n",
    "\n",
    "########### Execution Command ###################\n",
    "\n",
    "# Define directories\n",
    "INPUT_DIR=\"/scratch/MOryzae/SNP2/vcf_files\"\n",
    "OUTPUT_DIR=\"/scratch/MOryzae/SNP2/vcf_filtered\"\n",
    "\n",
    "# Load necessary modules\n",
    "module load bcftools/1.18\n",
    "module load vcftools/0.1.16\n",
    "module load htslib/1.19\n",
    "\n",
    "# set filters based on the summary statistics\n",
    "MAF=0.1                # Minor Allele Frequency minimal\n",
    "MISS=0.1               # Maximum missing data (3rd quartile)\n",
    "QUAL=15000                # Minimal variant quality\n",
    "\n",
    "# Input VCF file and output file names\n",
    "VCF_IN=\"${INPUT_DIR}/snp_correct.vcf.gz\"  # Input VCF file\n",
    "VCF_OUT=\"${OUTPUT_DIR}/filtered_snps.vcf.gz\"  # Output filtered VCF\n",
    "VCF_STATS=\"${OUTPUT_DIR}/filtered_snps_stats.txt\"  # File to save stats\n",
    "\n",
    "# move to the vcf directory\n",
    "cd $INPUT_DIR\n",
    "\n",
    "# Perform the filtering with vcftools\n",
    "vcftools --gzvcf $VCF_IN \\\n",
    "    --remove-indels \\\n",
    "    --maf $MAF \\\n",
    "    --max-missing $MISS \\\n",
    "    --minQ $QUAL \\\n",
    "    --min-meanDP $MIN_DEPTH \\\n",
    "    --max-meanDP $MAX_DEPTH \\\n",
    "      --recode \\\n",
    "    --stdout | bgzip -c > $VCF_OUT\n",
    "\n",
    "echo \"Filtrage terminé. Fichier filtré disponible à : $VCF_OUT\"\n",
    "\n",
    "# Index the filtered VCF file using bcftools\n",
    "echo \"Indexation du fichier filtré...\"\n",
    "bcftools index $VCF_OUT\n",
    "\n",
    "# Calculate the statistics of the filtered VCF\n",
    "echo \"Calcul des statistiques sur le fichier filtré...\"\n",
    "bcftools stats $VCF_OUT > $VCF_STATS\n",
    "\n",
    "echo \"Statistiques disponibles dans : $VCF_STATS\"\n",
    "echo \"Indexation terminée.\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "cf2008d8-3a08-41fc-b214-3729abf9fbe3",
   "metadata": {},
   "source": [
    "Points clés dans ce script :\n",
    "\n",
    "    Seuils basés sur vos statistiques :\n",
    "        QUAL < 999 pour correspondre au 1er quartile de la qualité.\n",
    "        MQ < 40 pour un seuil strict de la qualité de mapping.\n",
    "        --maf 0.05 pour exclure les variants rares (MAF < 5%).\n",
    "\n",
    "    Ajout des statistiques intermédiaires :\n",
    "        Les statistiques de qualité et après filtrage des SNPs sont enregistrées dans un fichier (filtering_stats.txt).\n",
    "\n",
    "    Optimisation de la gestion des fichiers :\n",
    "        Utilisation d’un fichier temporaire (temp_snps.vcf.gz) pour éviter d’écraser les résultats finaux avant vérification.\n",
    "\n",
    "    Filtres adaptés :\n",
    "        Conformité avec vos données : aucun SNP manquant ou profondeur calculée comme NA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "718250b1-4f10-49c9-82a9-ca0815ff55c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Lancer le script\n",
    "\n",
    "sbash snp2_filtering.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58870f88-2787-4da6-99ea-955ccdcb6028",
   "metadata": {},
   "source": [
    "## __Generate PCA using genotyping information contained in VCF__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a650caa-4cc8-4a93-b434-aea102cb0925",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Créer un répertoire PLINK dans le repertoire de travail\n",
    "\n",
    "mkdir -p /scratch/MOryzae/PLINK2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39e88ee1-f83e-40e4-b14b-af8118a6f23b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Se déplacer dans le répertoire créé\n",
    "\n",
    "cd PLINK2/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86077377-68aa-44f9-9941-1d0f22e307d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Changer de version de l'outil Plink pour des soucis d'options\n",
    "\n",
    "module load plink/1.9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8962a0bc-9263-4829-a281-19beabf13d87",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Lancer le run\n",
    "\n",
    "plink -vcf /scratch/MOryzae/SNP2/vcf_filtered/filtered_snps.vcf.gz --allow-extra-chr --cluster --matrix --pca 3 --mind --out ./plink/dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f390c024-1cbc-4744-9e33-e99db8c37177",
   "metadata": {},
   "source": [
    "## __Convertissons le fichier \"eigenvec\" généré en format csv__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77564018-f362-42f7-bb13-7e2816f09d39",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Se déplacer dans le répertoire créé\n",
    "\n",
    "cd /scratch/MOryzae/SCRIPTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da82ceb6-ef12-4aef-8017-4f2d24cc41fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Ouvrir l'éditeur de texte nano\n",
    "\n",
    "nano eigenvec2_snp_to_csv.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16e9462e-6cbf-4d83-b5ef-6fb97fc35166",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############# SLURM Configuration ##############\n",
    "\n",
    "### Define Job name\n",
    "#SBATCH --job-name=eigenvec_to_csv\n",
    "\n",
    "### Define partition to use\n",
    "#SBATCH -p normal\n",
    "\n",
    "### Define number of CPUs to use\n",
    "#SBATCH -c 8\n",
    "\n",
    "### Specify the node to run on\n",
    "#SBATCH --nodelist=node20  # Run the job on node20\n",
    "\n",
    "#################################################\n",
    "\n",
    "########### Execution Command ###################\n",
    "\n",
    "# Define directories\n",
    "INPUT_DIR=\"/scratch/MOryzae/PLINK2\"\n",
    "OUTPUT_DIR=\"/scratch/MOryzae/PLINK2\"\n",
    "\n",
    "# List of directories to process\n",
    "DIRECTORIES=(\"plink\")\n",
    "\n",
    "# Loop through each directory\n",
    "for DIR in \"${DIRECTORIES[@]}\"; do\n",
    "    PCA_FILE=\"$INPUT_DIR/$DIR/dataset.eigenvec\"\n",
    "    OUTPUT_CSV=\"$OUTPUT_DIR/$DIR/dataset.csv\"\n",
    "\n",
    "    # Check if PCA results exist\n",
    "    if [ -f \"$PCA_FILE\" ]; then\n",
    "        echo \"Processing PCA results for $DIR...\"\n",
    "\n",
    "        # Convert eigenvec file to CSV format\n",
    "        awk 'NR==1{print \"FID,IID,PC1,PC2,PC3\"} NR>1{print $1\",\"$2\",\"$3\",\"$4\",\"$5}' \"$PCA_FILE\" > \"$OUTPUT_CSV\"\n",
    "\n",
    "        # Check if conversion was successful\n",
    "        if [ $? -eq 0 ]; then\n",
    "            echo \"Conversion successful: $OUTPUT_CSV created.\"\n",
    "        else\n",
    "            echo \"Error: Failed to convert $PCA_FILE to CSV.\"\n",
    "            exit 1\n",
    "        fi\n",
    "    else\n",
    "        echo \"Error: PCA results file not found in $DIR.\"\n",
    "        exit 1\n",
    "    fi\n",
    "done\n",
    "\n",
    "echo \"PCA analysis completed. Results are saved in $OUTPUT_DIR.\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "624df533-fa39-41de-aa1f-5ba70756bdef",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Lancer le script\n",
    "\n",
    "sbash eigenvec2_snp_to_csv.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa52d5fa-36ff-4002-a294-5b8844db1172",
   "metadata": {},
   "source": [
    "## __Sortir un plot pour la PCA afin de faciliter la visualisation__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e814e709-3815-44ae-9387-d27b7660e2b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Ouvrir l'éditeur de texte nano\n",
    "\n",
    "nano pca2_snp_plot.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1f92923-f122-4445-907e-a0da90510dda",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############# SLURM Configuration ##############\n",
    "\n",
    "### Define Job name\n",
    "#SBATCH --job-name=genome_pca_plot\n",
    "\n",
    "### Define partition to use\n",
    "#SBATCH -p normal\n",
    "\n",
    "### Define number of CPUs to use\n",
    "#SBATCH -c 8\n",
    "\n",
    "### Specify the node to run on\n",
    "#SBATCH --nodelist=node20\n",
    "\n",
    "#################################################\n",
    "\n",
    "########### Execution Command ###################\n",
    "\n",
    "module load python/3.12.0  # Charge Python 3.12 sur le cluster\n",
    "\n",
    "# Define directories\n",
    "PCA_RESULTS_DIR=\"/scratch/MOryzae/PLINK2\"\n",
    "OUTPUT_PLOT_DIR=\"/scratch/MOryzae/PLINK2\"\n",
    "\n",
    "# List of directories to process\n",
    "DIRECTORIES=(\"plink\")\n",
    "\n",
    "# Loop through each directory\n",
    "for DIRECTORY in \"${DIRECTORIES[@]}\"; do\n",
    "    PCA_FILE=\"$PCA_RESULTS_DIR/$DIRECTORY/dataset.eigenvec\"\n",
    "    OUTPUT_DIR=\"$OUTPUT_PLOT_DIR/$DIRECTORY\"\n",
    "    OUTPUT_PLOT_2D=\"$OUTPUT_DIR/dataset_2D.png\"\n",
    "    OUTPUT_PLOT_3D=\"$OUTPUT_DIR/dataset_3D.png\"\n",
    "\n",
    "    # Ensure the output directory exists\n",
    "    mkdir -p \"$OUTPUT_DIR\"\n",
    "\n",
    "    # Check if the PCA results file exists\n",
    "    if [ -f \"$PCA_FILE\" ]; then\n",
    "        echo \"Processing PCA results for $DIRECTORY...\"\n",
    "        \n",
    "        # Call the Python script to generate the plots\n",
    "        python3 <<EOF\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "import os\n",
    "\n",
    "# Define input and output paths\n",
    "pca_results_file = \"$PCA_FILE\"\n",
    "output_plot_2D = \"$OUTPUT_PLOT_2D\"\n",
    "output_plot_3D = \"$OUTPUT_PLOT_3D\"\n",
    "\n",
    "# Read PCA results\n",
    "try:\n",
    "    pca_results = pd.read_csv(pca_results_file, sep=r'\\s+', header=None)\n",
    "    pca_results.columns = ['FID', 'IID', 'PC1', 'PC2', 'PC3']\n",
    "except Exception as e:\n",
    "    print(f\"Error reading PCA results file {pca_results_file}: {e}\")\n",
    "    exit(1)\n",
    "\n",
    "# Plot 2D scatter plot for PC1 vs PC2\n",
    "try:\n",
    "    plt.figure(figsize=(8, 6))\n",
    "    plt.scatter(pca_results['PC1'], pca_results['PC2'], s=100)\n",
    "    plt.title('PCA Results: $DIRECTORY (2D)')\n",
    "    plt.xlabel('Principal Component 1')\n",
    "    plt.ylabel('Principal Component 2')\n",
    "    plt.grid()\n",
    "    plt.savefig(output_plot_2D)\n",
    "    plt.close()\n",
    "except Exception as e:\n",
    "    print(f\"Error creating 2D plot for {pca_results_file}: {e}\")\n",
    "    exit(1)\n",
    "\n",
    "# Plot 3D scatter plot for PC1, PC2, and PC3\n",
    "try:\n",
    "    fig = plt.figure(figsize=(10, 8))\n",
    "    ax = fig.add_subplot(111, projection='3d')\n",
    "    scatter = ax.scatter(pca_results['PC1'], pca_results['PC2'], pca_results['PC3'], s=100, c='blue', alpha=0.7)\n",
    "    ax.set_title('PCA Results: $DIRECTORY (3D)')\n",
    "    ax.set_xlabel('Principal Component 1')\n",
    "    ax.set_ylabel('Principal Component 2')\n",
    "    ax.set_zlabel('Principal Component 3')\n",
    "    plt.savefig(output_plot_3D)\n",
    "    plt.close()\n",
    "except Exception as e:\n",
    "    print(f\"Error creating 3D plot for {pca_results_file}: {e}\")\n",
    "    exit(1)\n",
    "\n",
    "# Verify plots were created\n",
    "if not os.path.exists(output_plot_2D) or not os.path.exists(output_plot_3D):\n",
    "    print(f\"Error: Output plots not created for {pca_results_file}\")\n",
    "    exit(1)\n",
    "EOF\n",
    "\n",
    "        # Check if the Python script executed successfully\n",
    "        if [ $? -eq 0 ]; then\n",
    "            echo \"Plots successfully created for $DIRECTORY: $OUTPUT_PLOT_2D, $OUTPUT_PLOT_3D\"\n",
    "        else\n",
    "            echo \"Error: Failed to create plots for $DIRECTORY.\"\n",
    "            exit 1\n",
    "        fi\n",
    "    else\n",
    "        echo \"Error: PCA results file not found in $DIRECTORY.\"\n",
    "        exit 1\n",
    "    fi\n",
    "done\n",
    "\n",
    "echo \"All PCA plots created successfully.\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77db72c4-a853-46bb-a2a6-fa4d5f51c9ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Lancer le script\n",
    "\n",
    "sbash pca2_snp_plot.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9da4fa58-ec16-4e75-865b-1572075dfc94",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Se déplacer dans le répertoire créé\n",
    "\n",
    "cd /scratch/MOryzae/SCRIPTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce0f1e75-86f9-44eb-bad2-27927a196584",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Ouvrir l'éditeur de texte nano\n",
    "\n",
    "nano pca2_snp_learning.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae4f9d0c-4789-4e4d-be4a-011b7a879907",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############# SLURM Configuration ##############\n",
    "#SBATCH --job-name=pca_learning\n",
    "#SBATCH -p normal\n",
    "#SBATCH -c 8\n",
    "#SBATCH --nodelist=node20\n",
    "\n",
    "#################################################\n",
    "\n",
    "# Load necessary modules\n",
    "module load python/3.12.0\n",
    "\n",
    "# Define directories\n",
    "PCA_RESULTS_DIR=\"/scratch/MOryzae/PLINK2\"\n",
    "OUTPUT_PLOT_DIR=\"/scratch/MOryzae/PLINK2/PCA\"\n",
    "\n",
    "# Create output directory if it doesn't exist\n",
    "mkdir -p \"$OUTPUT_PLOT_DIR\"\n",
    "\n",
    "# List of subdirectories to process\n",
    "DIRECTORIES=(\"plink\")\n",
    "\n",
    "# Loop over each directory\n",
    "for DIRECTORY in \"${DIRECTORIES[@]}\"; do\n",
    "    PCA_FILE=\"$PCA_RESULTS_DIR/$DIRECTORY/dataset.eigenvec\"\n",
    "    OUTPUT_DIR=\"$OUTPUT_PLOT_DIR\"\n",
    "\n",
    "    # Check if the PCA file exists\n",
    "    if [[ -f \"$PCA_FILE\" ]]; then\n",
    "        mkdir -p \"$OUTPUT_DIR\"\n",
    "        echo \"Processing PCA file: $PCA_FILE\"\n",
    "\n",
    "        # Run the Python script for clustering and plotting\n",
    "        python3 clustering_analysis.py \"$PCA_FILE\" \"$OUTPUT_DIR\"\n",
    "\n",
    "        echo \"Processing completed for directory: $DIRECTORY\"\n",
    "    else\n",
    "        echo \"PCA file not found: $PCA_FILE\"\n",
    "    fi\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc82e437-216c-4910-abb2-5c8f34f59a1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Lancer le script\n",
    "\n",
    "sbash pca2_snp_learning.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd82f962-8642-4a0c-9c14-598800d411f0",
   "metadata": {},
   "source": [
    "## __DAPC__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ffe14e2-e9e1-487a-a12b-5e64f3053ce1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Ouvrir l'éditeur de texte nano\n",
    "\n",
    "nano dapc2_snp_analysis.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d4d871f-e940-4de5-96b1-46ded78d62ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Charger les bibliothèques nécessaires\n",
    "if (!requireNamespace(\"adegenet\")) install.packages(\"adegenet\")\n",
    "if (!requireNamespace(\"factoextra\")) install.packages(\"factoextra\") # Pour le clustering\n",
    "if (!requireNamespace(\"ggplot2\")) install.packages(\"ggplot2\")\n",
    "if (!requireNamespace(\"dplyr\")) install.packages(\"dplyr\")\n",
    "\n",
    "library(adegenet)\n",
    "library(factoextra)\n",
    "library(ggplot2)\n",
    "library(dplyr)\n",
    "\n",
    "# Définir les chemins d'entrée et de sortie\n",
    "input_file <- \"/home/name/Documents/Projet_CIBiG/Mentoring_Project/Results/PLINK2/plink/dataset.eigenvec\"\n",
    "output_dir <- \"/home/name/Documents/Projet_CIBiG/Mentoring_Project/Results/PLINK2/DAPC/\"\n",
    "\n",
    "# Créer le répertoire de sortie si nécessaire\n",
    "if (!dir.exists(output_dir)) {\n",
    "  dir.create(output_dir, recursive = TRUE)\n",
    "  cat(\"Répertoire de sortie créé :\", output_dir, \"\\n\")\n",
    "}\n",
    "\n",
    "# Charger les données PCA\n",
    "cat(\"Chargement des données PCA...\\n\")\n",
    "pca_data <- read.table(input_file, header = FALSE)\n",
    "colnames(pca_data) <- c(\"FID\", \"IID\", \"PC1\", \"PC2\", \"PC3\")  # Modifier selon vos colonnes\n",
    "X <- pca_data %>% select(PC1, PC2, PC3)\n",
    "\n",
    "# Étape 1 : Clustering des individus (k-means)\n",
    "cat(\"Application de k-means pour générer des groupes...\\n\")\n",
    "set.seed(42)  # Pour assurer la reproductibilité\n",
    "n_clusters <- 3  # Ajustez selon vos besoins ou utilisez la méthode du coude (voir ci-dessous)\n",
    "kmeans_result <- kmeans(X, centers = n_clusters)\n",
    "\n",
    "# Ajouter les groupes au jeu de données\n",
    "pca_data$Group <- as.factor(kmeans_result$cluster)\n",
    "\n",
    "# Sauvegarder les groupes dans un fichier CSV\n",
    "groups_file <- file.path(output_dir, \"groups_kmeans.csv\")\n",
    "write.csv(pca_data, groups_file, row.names = FALSE)\n",
    "cat(\"Groupes sauvegardés dans :\", groups_file, \"\\n\")\n",
    "\n",
    "# Étape optionnelle : Déterminer le nombre optimal de clusters\n",
    "# Méthode du coude\n",
    "cat(\"Déterminer le nombre optimal de clusters avec la méthode du coude...\\n\")\n",
    "fviz_nbclust(X, kmeans, method = \"wss\") +\n",
    "  labs(title = \"Méthode du coude pour déterminer k\")\n",
    "\n",
    "# Étape 2 : DAPC\n",
    "cat(\"Optimisation du nombre de PCs pour DAPC...\\n\")\n",
    "dapc_initial <- dapc(X, pca_data$Group)\n",
    "optimal_pcs <- optim.a.score(dapc_initial)\n",
    "n_pcs <- optimal_pcs$n.pca\n",
    "cat(paste(\"Nombre optimal de PCs :\", n_pcs, \"\\n\"))\n",
    "\n",
    "# Réaliser la DAPC avec le nombre optimal de PCs\n",
    "dapc_result <- dapc(X, pca_data$Group, n.pca = n_pcs)\n",
    "\n",
    "# Étape 3 : Visualisation des résultats\n",
    "cat(\"Génération du graphique des clusters DAPC...\\n\")\n",
    "scatter_file <- file.path(output_dir, \"dapc_scatter.png\")\n",
    "png(scatter_file, width = 800, height = 600)\n",
    "scatter(dapc_result, scree.da = TRUE, posi.da = \"bottomleft\", scree.pca = TRUE)\n",
    "dev.off()\n",
    "cat(\"Graphique DAPC sauvegardé dans :\", scatter_file, \"\\n\")\n",
    "\n",
    "# Graphique ggplot des clusters\n",
    "dapc_df <- data.frame(dapc_result$ind.coord) %>%\n",
    "  mutate(Group = pca_data$Group)  # Ajouter les groupes au dataframe\n",
    "\n",
    "ggplot_file <- file.path(output_dir, \"dapc_ggplot.png\")\n",
    "gg <- ggplot(dapc_df, aes(x = LD1, y = LD2, color = Group)) +\n",
    "  geom_point(size = 3, alpha = 0.8) +\n",
    "  theme_minimal() +\n",
    "  labs(title = \"Clusters DAPC\", x = \"Discriminant Axis 1\", y = \"Discriminant Axis 2\")\n",
    "ggsave(ggplot_file, plot = gg, width = 8, height = 6)\n",
    "cat(\"Graphique ggplot DAPC sauvegardé dans :\", ggplot_file, \"\\n\")\n",
    "\n",
    "# Étape 4 : Contributions des variables\n",
    "cat(\"Visualisation des contributions des variables...\\n\")\n",
    "loading_file <- file.path(output_dir, \"dapc_loadings.png\")\n",
    "png(loading_file, width = 800, height = 600)\n",
    "loadingplot(dapc_result$var.contr, axis = 1, threshold = 0.005, lab.jitter = 1)\n",
    "dev.off()\n",
    "cat(\"Graphique des contributions sauvegardé dans :\", loading_file, \"\\n\")\n",
    "\n",
    "# Étape 5 : Sauvegarder les résultats\n",
    "results_file <- file.path(output_dir, \"dapc_results_with_clusters.csv\")\n",
    "write.csv(dapc_df, results_file, row.names = FALSE)\n",
    "cat(\"Résultats DAPC sauvegardés dans :\", results_file, \"\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a032a298-03f9-4213-954d-36307e5c80a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Lancer le script sur votre machine locale dans R\n",
    "\n",
    "source(\"/path/to/working dorectory/on your laptop/dapc2_snp_analysis.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e60ea7c8-5af8-41bc-bb52-d6b9fbb63f4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Copier les données vers le NAS\n",
    "\n",
    "scp -r /scratch/MOryzae/PLINK2/ san:/projects/medium/CIBiG_MOryzae/"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
