{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fd3faf54-4921-4e7e-9859-1fd8d5133805",
   "metadata": {},
   "source": [
    "# __STRATEGIE D'ANALYSE BIOINFORMATIQUE - MENTORING PROJECT__\n",
    "\n",
    "__AGMIMONHAN Attolou Raoul, NAME Pakyendou Estel__\n",
    "\n",
    "__Tuteurs: Aurore COMTE & Sebastien RAVEL__\n",
    "\n",
    "Jupyter inspired by the model created by C. Tranchant (DIADE-IRD), J. Orjuela (DIADE-IRD), F. Sabot (DIADE-IRD) and A. Dereeper (PHIM-IRD)\n",
    "***\n",
    "\n",
    "# <span style=\"color: #006E7F\">Table of contents</span>\n",
    "<a class=\"anchor\" id=\"home\"></a>\n",
    "\n",
    "\n",
    "[PRACTICE III - Mapping sur tous les échantillons](#mapping) \n",
    "   * [Reference indexation  `bwa-mem2 index`](#refindex)\n",
    "   * [Mapping avec `bwa-mem2 mem`](#bwamem2-cmd)\n",
    "   * [Convertir sam into bam `samtools view`](#samtoolsview)\n",
    "   * [Etablir les statistiques du mapping `samtools flagstat`](#flagstats)\n",
    "   * [Filtrer les reads correctement mappés `samtools view`](#corrmap)\n",
    "   * [Faire le sort des reads filtrés `samtools sort`](#sort)\n",
    "   * [Indexation des fichiers bam sorted  `samtools index`](#indexbam) \n",
    "   \n",
    "\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5706c2ff-7a31-4137-a094-efe60d115e29",
   "metadata": {},
   "source": [
    "# __Practice III - Mapping sur tous les échantillons__"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f709314-ae60-423e-a300-edf130ebc09d",
   "metadata": {},
   "source": [
    "## __III.1. Téléchargement & Indexation du génome de référence__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc8c3678-ae86-4e9c-a65c-cdc8465ae965",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Se déplacer dans le répertoire SCRIPTS\n",
    "\n",
    "cd /scratch/MOryzae/SCRIPTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78247a42-78f6-4829-b97c-dd149c34f2f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Ouvrir l'éditeur de texte nano\n",
    "\n",
    "nano Refseq.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5684604-6237-4840-9eea-2a7d5fef70c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############# SLURM Configuration ##############\n",
    "\n",
    "### Define Job name\n",
    "#SBATCH --job-name=genome_download_index\n",
    "\n",
    "### Define partition to use\n",
    "#SBATCH -p normal\n",
    "\n",
    "### Define number of CPUs to use\n",
    "#SBATCH -c 8\n",
    "\n",
    "### Specify the node to run on\n",
    "#SBATCH --nodelist=node20  # Spécifie que le job doit être exécuté sur node20\n",
    "\n",
    "#################################################\n",
    "\n",
    "########### Execution Command ###################\n",
    "\n",
    "# Créer le répertoire REF si nécessaire\n",
    "REF_DIR=\"/scratch/MOryzae/REF\"\n",
    "mkdir -p \"$REF_DIR\"\n",
    "\n",
    "# Définir le chemin du fichier de référence\n",
    "GENOME_URL=\"https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/495/GCF_000002495.2_MG8/GCF_000002495.2_MG8_genomic.fna.gz\"\n",
    "GENOME_FILE=\"$REF_DIR/MOryzae_genomic.fna.gz\"\n",
    "\n",
    "# Télécharger le génome de référence\n",
    "echo \"Téléchargement du génome de référence...\"\n",
    "wget -O \"$GENOME_FILE\" \"$GENOME_URL\"\n",
    "\n",
    "# Vérifier si le téléchargement a réussi\n",
    "if [[ $? -ne 0 ]]; then\n",
    "    echo \"Erreur : Le téléchargement du génome a échoué.\"\n",
    "    exit 1\n",
    "fi\n",
    "\n",
    "# Décompresser le fichier\n",
    "echo \"Décompression du fichier...\"\n",
    "gunzip \"$GENOME_FILE\"\n",
    "\n",
    "# Modifier le nom pour une utilisation plus facile\n",
    "mv \"${GENOME_FILE%.gz}\" \"$REF_DIR/MOryzae_genomic.fna\"\n",
    "\n",
    "# Charger le module bwa-mem2 pour faire l'indexation\n",
    "module load bwamem2/2.2.1\n",
    "\n",
    "# Indexer le génome de référence\n",
    "echo \"Indexation du génome de référence...\"\n",
    "bwa-mem2 index \"$REF_DIR/MOryzae_genomic.fna\"\n",
    "\n",
    "# Vérifier si l'indexation a réussi\n",
    "if [[ $? -eq 0 ]]; then\n",
    "    echo \"Indexation du génome de référence réussie.\"\n",
    "else\n",
    "    echo \"Erreur : L'indexation du génome a échoué.\"\n",
    "fi\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "940f00f1-59cc-47c1-89b5-752193e440b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Lancer le script\n",
    "\n",
    "sbash refseq.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c2de8e1-4281-41fb-a6a1-163f3416aa63",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Vérifier les fichiers index générés et leurs extensions\n",
    "\n",
    "ls -lh /scratch/MOryzae/REF/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f27b438-beaf-414c-9166-094c0f429567",
   "metadata": {},
   "source": [
    "## __III.2. Mapping__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b70eb8e-f5aa-4bb3-ab0f-d82a92487255",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Créer le répertoire MAPPING dans le répertoire de travail\n",
    "\n",
    "mkdir -p /scratch/MOryzae/MAPPING\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44d45fe2-2023-497c-95ac-0b87e5df76b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Ouvrir l'éditeur de texte nano\n",
    "\n",
    "nano mapping_pipeline.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e422b08-5ba5-41d7-9dbe-a245949f60f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############# SLURM Configuration ##############\n",
    "\n",
    "### Set the job name\n",
    "#SBATCH --job-name=mapping_pipeline\n",
    "\n",
    "### Set the partition to use\n",
    "#SBATCH -p normal\n",
    "\n",
    "### Set the number of CPUs to use\n",
    "#SBATCH -c 8\n",
    "\n",
    "### Specify the node on which the job should run\n",
    "#SBATCH --nodelist=node20  # Specifies that the job should run on node20\n",
    "\n",
    "#################################################\n",
    "\n",
    "########### Path Variables ##################\n",
    "\n",
    "# Define paths for data and output directories\n",
    "REF_PATH=\"/scratch/MOryzae/REF/MOryzae_genomic.fna\"\n",
    "TRIM_DATA_PATH=\"/scratch/MOryzae/DATA/Trimming\"\n",
    "OUTPUT_PATH=\"/scratch/MOryzae/MAPPING\"\n",
    "SAM_PATH=\"${OUTPUT_PATH}/sam_files\"\n",
    "BAM_PATH=\"${OUTPUT_PATH}/bam_raw\"\n",
    "STATS_PATH=\"${OUTPUT_PATH}/bam_stats\"\n",
    "FILTERED_PATH=\"${OUTPUT_PATH}/bam_filtered\"\n",
    "SORTED_PATH=\"${OUTPUT_PATH}/bam_mapped_sort\"\n",
    "\n",
    "# Load bwa-mem2 and samtools modules\n",
    "module load bwamem2/2.2.1\n",
    "module load samtools/1.18\n",
    "\n",
    "# List of sequences\n",
    "sequences=(\"AG0004\" \"BN0123\" \"CH0461\" \"G22\" \"IE1K\" \"IR0015\" \"ML33\" \"PH42\" \"TN0057\"\n",
    "           \"Arcadia\" \"BN0202\" \"CH0533\" \"GFSI1-7-2\" \"IN0017\" \"IR0083\" \"NG0012\" \"PL2-1\" \"TN0065\"\n",
    "           \"B2\" \"BN0252\" \"CH1103\" \"GG11\" \"IN0054\" \"IR0084\" \"NG0054\" \"SSFL02\" \"TN0090\"\n",
    "           \"B71\" \"Br7\" \"CH1164\" \"GN0001\" \"IN0059\" \"IR0088\" \"NP0058\" \"SSFL14-3\" \"TR0025\"\n",
    "           \"Bd8401\" \"Br80\" \"CHRF\" \"GY0040\" \"IN0114\" \"IR0095\" \"P28\" \"T25\" \"US0041\"\n",
    "           \"BdBar\" \"CD0065\" \"CHW\" \"HO\" \"IN0115\" \"IT0010\" \"P29\" \"TG0004\" \"US0064\"\n",
    "           \"BF0072\" \"CD0142\" \"CM0028\" \"IA1\" \"IN0116\" \"JP0091\" \"P3\" \"TG0032\" \"VT0027\"\n",
    "           \"Bm88324\" \"CH0043\" \"FR1067\" \"IB33\" \"INA168\" \"LpKY-97-1\" \"Pg1213-22\" \"TN0001\" \"VT0030\"\n",
    "           \"BN0019\" \"CH0072\" \"FR1069\" \"IB49\" \"IR00102\" \"ML0060\" \"PgKY4OV2-1\" \"TN0002\" \"Z2-1\"\n",
    "           \"BN0119\" \"CH0452\" \"G17\" \"IC17\" \"IR0013\" \"ML0062\" \"PgPA18C-02\" \"TN0050\")\n",
    "\n",
    "#################################################\n",
    "\n",
    "########### Create output directories ##################\n",
    "\n",
    "# Create the output directories if they do not exist\n",
    "mkdir -p \"$SAM_PATH\" \"$BAM_PATH\" \"$STATS_PATH\" \"$FILTERED_PATH\" \"$SORTED_PATH\"\n",
    "\n",
    "########### Step 1: Mapping, SAM to BAM conversion, Statistics, Filtering, Sorting ##################\n",
    "\n",
    "\n",
    "# Loop over each sequence to perform all steps\n",
    "for sequence in \"${sequences[@]}\"; do\n",
    "    echo -e \"######################\\nProcessing for ${sequence}...\"\n",
    "    \n",
    "    # Define file paths for input and output\n",
    "    R1=\"${TRIM_DATA_PATH}/${sequence}_R1_paired.fastq.gz\"\n",
    "    R2=\"${TRIM_DATA_PATH}/${sequence}_R2_paired.fastq.gz\"\n",
    "    SAM_FILE=\"${SAM_PATH}/${sequence}.sam\"\n",
    "    BAM_FILE=\"${BAM_PATH}/${sequence}.bam\"\n",
    "    FLAGSTAT_FILE=\"${STATS_PATH}/${sequence}.flagstat\"\n",
    "    FILTERED_BAM=\"${FILTERED_PATH}/${sequence}.mappedpaired.bam\"\n",
    "    SORTED_BAM=\"${SORTED_PATH}/${sequence}.mappedpaired.sorted.bam\"\n",
    "    \n",
    "    # Step 1: Mapping with bwa-mem2\n",
    "    bwa-mem2 mem -t 8 \"$REF_PATH\" \"$R1\" \"$R2\" -o \"$SAM_FILE\"\n",
    "    echo \"Mapping completed for ${sequence}\"\n",
    "    \n",
    "    # Step 2: Convert SAM to BAM\n",
    "    samtools view -b -o \"$BAM_FILE\" \"$SAM_FILE\"\n",
    "    echo \"SAM to BAM conversion successful for ${sequence}\"\n",
    "    \n",
    "    # Step 3: Generate statistics using flagstat\n",
    "    samtools flagstat -@ 8 \"$BAM_FILE\" > \"$FLAGSTAT_FILE\"\n",
    "    echo \"Statistics generated for ${sequence}: ${FLAGSTAT_FILE}\"\n",
    "\n",
    "    # Step 4: Filter BAM files\n",
    "    samtools view -bh -@ 8 -f 0x02 -o \"$FILTERED_BAM\" \"$BAM_FILE\"\n",
    "    echo \"Filtered BAM created for ${sequence}: ${FILTERED_BAM}\"\n",
    "    \n",
    "    # Step 5: Sort the filtered BAM files\n",
    "    samtools sort -@ 8 \"$FILTERED_BAM\" -o \"$SORTED_BAM\"\n",
    "    echo \"Sorted BAM created for ${sequence}: ${SORTED_BAM}\"\n",
    "    \n",
    "    # Step 6: Index the sorted BAM file\n",
    "    samtools index \"$SORTED_BAM\"\n",
    "    echo \"Indexing completed for ${sequence}\"\n",
    "    \n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56b8e4d2-0637-4884-9b6f-0a7ed5164d65",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Lancer le script\n",
    "\n",
    "sbash mapping_pipeline.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c889439-1546-478f-bf7c-31e857ae720a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Grouper le fichiers flagstaten un seul fichier csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94ff57c4-b5a4-4dbd-917c-7e7afc86dc95",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Ouvrir l'éditeur de texte nano\n",
    "\n",
    "nano flagstat.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcaae074-14c0-4086-94d9-39c9fb304eee",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############# SLURM Configuration ##############\n",
    "\n",
    "### Set the job name\n",
    "#SBATCH --job-name=concatenate_flagstat\n",
    "\n",
    "### Set the partition to use\n",
    "#SBATCH -p normal\n",
    "\n",
    "### Set the number of CPUs to use\n",
    "#SBATCH -c 8\n",
    "\n",
    "### Specify the node on which the job should run\n",
    "#SBATCH --nodelist=node20  # Specifies that the job should run on node20\n",
    "\n",
    "#################################################\n",
    "\n",
    "# Path variables\n",
    "flagstat_dir=\"/scratch/MOryzae/MAPPING/bam_stats\"  # Directory containing flagstat files\n",
    "stat_file=\"/scratch/MOryzae/MAPPING/bam_stats/all_stat.csv\"  # Output file\n",
    "\n",
    "# List of prefix\n",
    "prefix=(\"AG0004\" \"BN0123\" \"CH0461\" \"G22\" \"IE1K\" \"IR0015\" \"ML33\" \"PH42\" \"TN0057\"\n",
    "           \"Arcadia\" \"BN0202\" \"CH0533\" \"GFSI1-7-2\" \"IN0017\" \"IR0083\" \"NG0012\" \"PL2-1\" \"TN0065\"\n",
    "           \"B2\" \"BN0252\" \"CH1103\" \"GG11\" \"IN0054\" \"IR0084\" \"NG0054\" \"SSFL02\" \"TN0090\"\n",
    "           \"B71\" \"Br7\" \"CH1164\" \"GN0001\" \"IN0059\" \"IR0088\" \"NP0058\" \"SSFL14-3\" \"TR0025\"\n",
    "           \"Bd8401\" \"Br80\" \"CHRF\" \"GY0040\" \"IN0114\" \"IR0095\" \"P28\" \"T25\" \"US0041\"\n",
    "           \"BdBar\" \"CD0065\" \"CHW\" \"HO\" \"IN0115\" \"IT0010\" \"P29\" \"TG0004\" \"US0064\"\n",
    "           \"BF0072\" \"CD0142\" \"CM0028\" \"IA1\" \"IN0116\" \"JP0091\" \"P3\" \"TG0032\" \"VT0027\"\n",
    "           \"Bm88324\" \"CH0043\" \"FR1067\" \"IB33\" \"INA168\" \"LpKY-97-1\" \"Pg1213-22\" \"TN0001\" \"VT0030\"\n",
    "           \"BN0019\" \"CH0072\" \"FR1069\" \"IB49\" \"IR00102\" \"ML0060\" \"PgKY4OV2-1\" \"TN0002\" \"Z2-1\"\n",
    "           \"BN0119\" \"CH0452\" \"G17\" \"IC17\" \"IR0013\" \"ML0062\" \"PgPA18C-02\" \"TN0050\")\n",
    "\n",
    "# Create the output file and write the headers\n",
    "echo \"Sequence,total,primary,secondary,supplementary,duplicates,primary_duplicates,mapped,primary_mapped,paired_in_sequencing,read1,read2,properly_paired,singletons,unmapped\" > \"$stat_file\"\n",
    "\n",
    "# Loop through each sequence\n",
    "for file in \"${prefix[@]}\"; do\n",
    "    # Define the corresponding flagstat file for the sequence\n",
    "    flagstat_file=\"${flagstat_dir}/${file}.flagstat\"\n",
    "\n",
    "    # Check if the file exists\n",
    "    if [[ -f \"$flagstat_file\" ]]; then\n",
    "        # Initialize variables to store values from the flagstat file\n",
    "        total=0\n",
    "        primary=0\n",
    "        secondary=0\n",
    "        supplementary=0\n",
    "        duplicates=0\n",
    "        primary_duplicates=0\n",
    "        mapped=0\n",
    "        primary_mapped=0\n",
    "        paired_in_sequencing=0\n",
    "        read1=0\n",
    "        read2=0\n",
    "        properly_paired=0\n",
    "        singletons=0\n",
    "        unmapped=0\n",
    "\n",
    "        # Read the flagstat file line by line\n",
    "        while IFS= read -r line; do\n",
    "            # Extract values based on the specific keywords\n",
    "            if [[ $line =~ ^([0-9]+)\\ \\+\\ [0-9]+\\ in\\ total ]]; then\n",
    "                total=\"${BASH_REMATCH[1]}\"\n",
    "            elif [[ $line =~ ^([0-9]+)\\ \\+\\ [0-9]+\\ primary ]]; then\n",
    "                primary=\"${BASH_REMATCH[1]}\"\n",
    "            elif [[ $line =~ ^([0-9]+)\\ \\+\\ [0-9]+\\ secondary ]]; then\n",
    "                secondary=\"${BASH_REMATCH[1]}\"\n",
    "            elif [[ $line =~ ^([0-9]+)\\ \\+\\ [0-9]+\\ supplementary ]]; then\n",
    "                supplementary=\"${BASH_REMATCH[1]}\"\n",
    "            elif [[ $line =~ ^([0-9]+)\\ \\+\\ [0-9]+\\ duplicates ]]; then\n",
    "                duplicates=\"${BASH_REMATCH[1]}\"\n",
    "            elif [[ $line =~ ^([0-9]+)\\ \\+\\ [0-9]+\\ primary\\ duplicates ]]; then\n",
    "                primary_duplicates=\"${BASH_REMATCH[1]}\"\n",
    "            elif [[ $line =~ ^([0-9]+)\\ \\+\\ [0-9]+\\ mapped ]]; then\n",
    "                mapped=\"${BASH_REMATCH[1]}\"\n",
    "            elif [[ $line =~ ^([0-9]+)\\ \\+\\ [0-9]+\\ primary\\ mapped ]]; then\n",
    "                primary_mapped=\"${BASH_REMATCH[1]}\"\n",
    "            elif [[ $line =~ ^([0-9]+)\\ \\+\\ [0-9]+\\ paired\\ in\\ sequencing ]]; then\n",
    "                paired_in_sequencing=\"${BASH_REMATCH[1]}\"\n",
    "            elif [[ $line =~ ^([0-9]+)\\ \\+\\ [0-9]+\\ read1 ]]; then\n",
    "                read1=\"${BASH_REMATCH[1]}\"\n",
    "            elif [[ $line =~ ^([0-9]+)\\ \\+\\ [0-9]+\\ read2 ]]; then\n",
    "                read2=\"${BASH_REMATCH[1]}\"\n",
    "            elif [[ $line =~ ^([0-9]+)\\ \\+\\ [0-9]+\\ properly\\ paired ]]; then\n",
    "                properly_paired=\"${BASH_REMATCH[1]}\"\n",
    "            elif [[ $line =~ ^([0-9]+)\\ \\+\\ [0-9]+\\ singletons ]]; then\n",
    "                singletons=\"${BASH_REMATCH[1]}\"\n",
    "            fi\n",
    "        done < \"$flagstat_file\"\n",
    "\n",
    "        # Calculate unmapped reads\n",
    "        unmapped=$((total - mapped))\n",
    "\n",
    "        # Write the results to the CSV file\n",
    "        echo \"$file,$total,$primary,$secondary,$supplementary,$duplicates,$primary_duplicates,$mapped,$primary_mapped,$paired_in_sequencing,$read1,$read2,$properly_paired,$singletons,$unmapped\" >> \"$stat_file\"\n",
    "    else\n",
    "        echo \"Warning: File for sequence $seq not found. Skipping.\"\n",
    "    fi\n",
    "done\n",
    "\n",
    "echo \"Data extraction complete. Results saved in $stat_file.\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70824d74-9faa-48de-a89d-f4541d1bcf23",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Supprimer les sous répertoires sam_files, bam_raw, bam_filtered dans le répertoire MAPPING afin de libérer plus d'espaces\n",
    "\n",
    "rm -rf /scratch/MOryzae/MAPPING/sam_files\n",
    "rm -rf /scratch/MOryzae/MAPPING/bam_raw\n",
    "rm -rf /scratch/MOryzae/MAPPING/bam_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b9f33ff-cc91-4b8b-994a-4698aea90632",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Récupérer tous les outputs générés dans le répertoire MAPPING sur le NAS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99b9eda7-aca3-44fb-a378-e81c9cef1cba",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Déplacer tout le répertoire MAPPING vers le NAS\n",
    "\n",
    "scp -r /scratch/MOryzae/MAPPING san:/projects/medium/CIBiG_MOryzae/"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
